org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/embedder/MavenExecutionContext.java

```diff
@@ -147,19 +146,24 @@ protected MavenExecutionRequest newExecutionRequest() throws CoreException {
         throw new IllegalStateException(); //
       }
       request = DefaultMavenExecutionRequest.copy(parent);
+      request.setMultiModuleProjectDirectory(parent.getMultiModuleProjectDirectory());
     }
+    File multiModuleProjectDirectory = PlexusContainerManager.computeMultiModuleProjectDirectory(basedir);
     if(request == null) {
-      request = createExecutionRequest(IMavenConfiguration.getWorkspaceConfiguration(),
-          containerLookup, MavenPlugin.getMaven().getSettings());
+      request = createExecutionRequest(IMavenConfiguration.getWorkspaceConfiguration(), containerLookup,
+          MavenPlugin.getMaven().getSettings(), multiModuleProjectDirectory);
       request.setBaseDirectory(basedir);
+    } else if(!Objects.equals(multiModuleProjectDirectory, request.getMultiModuleProjectDirectory())) {
+      //TODO: this should have been removed, shouldn't it?
```

Comment:
I think this should be removed. If there are any obstacles, this is most likely because of a wrong call hierarchy and we should then fix this there.

Reply:
See this if still being hit, for example in:

```
Thread [Worker-8: Updating Maven Dependencies] (Suspended (breakpoint at line 159 in MavenExecutionContext))	
	MavenExecutionContext.newExecutionRequest() line: 159	
	MavenExecutionContext.getExecutionRequest() line: 135	
	ProjectRegistryManager.readMavenProjectFacades(Collection<IFile>, MutableProjectRegistry, IProgressMonitor) line: 742	
	ProjectRegistryManager.refresh(MutableProjectRegistry, DependencyResolutionContext, IProgressMonitor) line: 380	
	ProjectRegistryManager.refresh(MutableProjectRegistry, Collection<IFile>, IProgressMonitor) line: 360	
	ProjectRegistryRefreshJob.lambda$4(MutableProjectRegistry, Set, IMavenExecutionContext, IProgressMonitor) line: 154	
	0x0000000801941ef0.call(IMavenExecutionContext, IProgressMonitor) line: not available	
	MavenExecutionContext.executeBare(MavenProject, ICallable<V>, IProgressMonitor) line: 375	
	MavenExecutionContext.execute(MavenProject, ICallable<V>, IProgressMonitor) line: 279	
	MavenExecutionContext.execute(ICallable<V>, IProgressMonitor) line: 219	
	MavenImpl.execute(IMaven, boolean, boolean, ICallable<V>, IProgressMonitor) line: 1102	
	ProjectRegistryRefreshJob.lambda$0(MutableProjectRegistry, List, IProgressMonitor, IMavenExecutionContext, IProgressMonitor) line: 153	
	0x0000000801941cd0.call(IMavenExecutionContext, IProgressMonitor) line: not available	
	MavenExecutionContext.executeBare(MavenProject, ICallable<V>, IProgressMonitor) line: 375	
	MavenExecutionContext.execute(MavenProject, ICallable<V>, IProgressMonitor) line: 279	
	MavenExecutionContext.execute(ICallable<V>, IProgressMonitor) line: 219	
	ProjectRegistryRefreshJob.run(IProgressMonitor) line: 106	
	Worker.run() line: 63	
```

Reply:
But I'm still not sure if this is really so bad. Narrowing from a 'global' to a 'mvn-folder'-scoped context seems not too bad in general. But I have not yet fully thought this through.

Reply:
Actually it *is* bad, because that means you are probabbly executing with the wrong settings and/or container, `ProjectRegistryManager` should actually already group projects by their config/basedir.

Also think about that currently the "subcontext" modify the session from the outer one, but never restores this! 
So probably if we think this shoudl be supported, then we probably need to do it a bit earlier, where it is checked if a parent context exist and first test there if the parent is compatible, if yes, we can join, if not just start a new execution.

Reply:
> ProjectRegistryManager should actually already group projects by their config/basedir.

I'm not arguing about that, but if it does not work in this special case, isn't it then possible to not work in other cases as well? Should we all fix them? Therefore if possible I think it would be better to find a solution that hopefully works for all. 🤓 

> Also think about that currently the "subcontext" modify the session from the outer one, but never restores this!
So probably if we think this shoudl be supported, then we probably need to do it a bit earlier, where it is checked if a parent context exist and first test there if the parent is compatible, if yes, we can join, if not just start a new execution.

Indeed. Or could we create a new request, that can inherit values (it doesn't override) from the request of the parent?

Reply:
> Should we all fix them?

Absolutely! This is mandatory to fully support such advanced features, if you just use "anything" to execute your task your outcome will not be correct (as seen with the recent failure with the copy-resources) so even though it is some work I thin we cannot avoid it if we want things to work consistent.